# 1. Import libraries

In [2]:
# Importing necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plot
import seaborn as sns


# 2. Load data

In [3]:
# Load data
df_customer = pd.read_csv("../data/raw/olist_customers_dataset.csv")
df_geolocation = pd.read_csv("../data/raw/olist_geolocation_dataset.csv")
df_order_items = pd.read_csv("../data/raw/olist_order_items_dataset.csv")
df_order_payments = pd.read_csv("../data/raw/olist_order_payments_dataset.csv")
df_order_reviews = pd.read_csv("../data/raw/olist_order_reviews_dataset.csv")
df_orders = pd.read_csv("../data/raw/olist_orders_dataset.csv")
df_sellers = pd.read_csv("../data/raw/olist_sellers_dataset.csv")
df_products = pd.read_csv("../data/raw/olist_products_dataset.csv")
df_product_category_name_translation = pd.read_csv("../data/raw/product_category_name_translation.csv")

In [4]:
# Create a dictionary to hold all datasets for easy access
datasets = {
    "customers": df_customer,
    "geolocation": df_geolocation,
    "order_items": df_order_items,
    "order_payments": df_order_payments,
    "order_reviews": df_order_reviews,
    "orders": df_orders,
    "sellers": df_sellers,
    "products": df_products,
    "category_translation": df_product_category_name_translation,
}

# Function to profile a dataset
def profile_dataset(name, df):
    return {
        "---DATASET": name,
        "---ROWS": df.shape[0],
        "---COLUMNS": df.shape[1],
        "---COLS": df.columns.tolist(),
        "---MISSING_VALUES": df.isna().sum().sum(),
        "---DUPLICATES": df.duplicated().sum(),
        "---DATA_TYPES": df.dtypes.to_dict()
    }

# Function to find missing values per column
def missing_values_per_column(df, col_list):
    for col in col_list:
        print(f"Column: {col}")
        print(df[col].isna().sum())

# 3. Data Profiling
## 3.1. Customer data
##### Customer Profile

In [5]:
df_customer_profile = profile_dataset("customers", df_customer)
df_customer_profile

{'---DATASET': 'customers',
 '---ROWS': 99441,
 '---COLUMNS': 5,
 '---COLS': ['customer_id',
  'customer_unique_id',
  'customer_zip_code_prefix',
  'customer_city',
  'customer_state'],
 '---MISSING_VALUES': np.int64(0),
 '---DUPLICATES': np.int64(0),
 '---DATA_TYPES': {'customer_id': dtype('O'),
  'customer_unique_id': dtype('O'),
  'customer_zip_code_prefix': dtype('int64'),
  'customer_city': dtype('O'),
  'customer_state': dtype('O')}}

##### Data Glimpse

In [6]:
df_customer.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


Customer data seems clean, with no missing values and duplicates. It is generally string data with the zip code being the only numeric data.

## 3.2. Geolocation
##### Geolocation Profile

In [7]:
df_geolocation_profile = profile_dataset("geolocation", df_geolocation)
df_geolocation_profile

{'---DATASET': 'geolocation',
 '---ROWS': 1000163,
 '---COLUMNS': 5,
 '---COLS': ['geolocation_zip_code_prefix',
  'geolocation_lat',
  'geolocation_lng',
  'geolocation_city',
  'geolocation_state'],
 '---MISSING_VALUES': np.int64(0),
 '---DUPLICATES': np.int64(261831),
 '---DATA_TYPES': {'geolocation_zip_code_prefix': dtype('int64'),
  'geolocation_lat': dtype('float64'),
  'geolocation_lng': dtype('float64'),
  'geolocation_city': dtype('O'),
  'geolocation_state': dtype('O')}}

The geolocation data doesn't seem to have any missing values but it does have quite a lot of duplicated data. This data will need to be cleaned further.

##### Data Glimpse

In [8]:
df_geolocation.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


Looking at this data, the coordinate information is grouped by zip codes, which explains the huge amount of rows.

## 3.3. Order Items
##### Order Items Profile

In [9]:
df_order_items_profile = profile_dataset("order_items", df_order_items)
df_order_items_profile

{'---DATASET': 'order_items',
 '---ROWS': 112650,
 '---COLUMNS': 7,
 '---COLS': ['order_id',
  'order_item_id',
  'product_id',
  'seller_id',
  'shipping_limit_date',
  'price',
  'freight_value'],
 '---MISSING_VALUES': np.int64(0),
 '---DUPLICATES': np.int64(0),
 '---DATA_TYPES': {'order_id': dtype('O'),
  'order_item_id': dtype('int64'),
  'product_id': dtype('O'),
  'seller_id': dtype('O'),
  'shipping_limit_date': dtype('O'),
  'price': dtype('float64'),
  'freight_value': dtype('float64')}}

##### Data Glimpse

In [10]:
df_order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


The order items dataset semms clean, with no missing values and duplicates, and correct data types.

## 3.4. Order Payments
##### Profile

In [11]:
df_order_payments_profile = profile_dataset("order_payments", df_order_payments)
df_order_payments_profile

{'---DATASET': 'order_payments',
 '---ROWS': 103886,
 '---COLUMNS': 5,
 '---COLS': ['order_id',
  'payment_sequential',
  'payment_type',
  'payment_installments',
  'payment_value'],
 '---MISSING_VALUES': np.int64(0),
 '---DUPLICATES': np.int64(0),
 '---DATA_TYPES': {'order_id': dtype('O'),
  'payment_sequential': dtype('int64'),
  'payment_type': dtype('O'),
  'payment_installments': dtype('int64'),
  'payment_value': dtype('float64')}}

##### Data Glimpse

In [12]:
df_order_payments.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


The order payments dataset semms clean, with no missing values and duplicates, and correct data types.

## 3.5. Order Reviews
##### Profile

In [13]:
df_order_reviews_profile = profile_dataset("order_reviews", df_order_reviews)
df_order_reviews_profile

{'---DATASET': 'order_reviews',
 '---ROWS': 99224,
 '---COLUMNS': 7,
 '---COLS': ['review_id',
  'order_id',
  'review_score',
  'review_comment_title',
  'review_comment_message',
  'review_creation_date',
  'review_answer_timestamp'],
 '---MISSING_VALUES': np.int64(145903),
 '---DUPLICATES': np.int64(0),
 '---DATA_TYPES': {'review_id': dtype('O'),
  'order_id': dtype('O'),
  'review_score': dtype('int64'),
  'review_comment_title': dtype('O'),
  'review_comment_message': dtype('O'),
  'review_creation_date': dtype('O'),
  'review_answer_timestamp': dtype('O')}}

In [14]:
order_review_cols = df_order_reviews.columns.tolist()

missing_values_per_column(df_order_reviews, order_review_cols)

Column: review_id
0
Column: order_id
0
Column: review_score
0
Column: review_comment_title
87656
Column: review_comment_message
58247
Column: review_creation_date
0
Column: review_answer_timestamp
0


It seems the reviews dataset ontains missing values. However, the missing values are the review information itself, which won't always have a value. So, these missing values are not an actual problem.

In [15]:
df_order_reviews.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


## 3.6. Orders
### Profile

In [16]:
df_order_profile = profile_dataset("orders", df_orders)
df_order_profile

{'---DATASET': 'orders',
 '---ROWS': 99441,
 '---COLUMNS': 8,
 '---COLS': ['order_id',
  'customer_id',
  'order_status',
  'order_purchase_timestamp',
  'order_approved_at',
  'order_delivered_carrier_date',
  'order_delivered_customer_date',
  'order_estimated_delivery_date'],
 '---MISSING_VALUES': np.int64(4908),
 '---DUPLICATES': np.int64(0),
 '---DATA_TYPES': {'order_id': dtype('O'),
  'customer_id': dtype('O'),
  'order_status': dtype('O'),
  'order_purchase_timestamp': dtype('O'),
  'order_approved_at': dtype('O'),
  'order_delivered_carrier_date': dtype('O'),
  'order_delivered_customer_date': dtype('O'),
  'order_estimated_delivery_date': dtype('O')}}

In [17]:
order_cols = df_orders.columns.tolist()
missing_values_per_column(df_orders, order_cols)

Column: order_id
0
Column: customer_id
0
Column: order_status
0
Column: order_purchase_timestamp
0
Column: order_approved_at
160
Column: order_delivered_carrier_date
1783
Column: order_delivered_customer_date
2965
Column: order_estimated_delivery_date
0


It seems there are multiple columns with missing values. We can check each one to determine what they mean. The easiest first step is to see each one's `order_status`.

##### `order_approved_at`.

In [18]:
# Analyze the order status for orders that have not been approved
df_orders["order_status"][df_orders["order_approved_at"].isna()].value_counts()

order_status
canceled     141
delivered     14
created        5
Name: count, dtype: int64

`order_approved_at` has missing values across all three statuses. 141 orders were missing aproval time due to being cancelled and 5 were due to just being created. But there are 14 orders that were already delivered yet were missing approval time. Anomalies. Could be caused by other things, such as internal operations.

##### `order_delivered_carrier_date`

In [19]:
df_orders["order_status"][df_orders["order_delivered_carrier_date"].isna()].value_counts()

order_status
unavailable    609
canceled       550
invoiced       314
processing     301
created          5
approved         2
delivered        2
Name: count, dtype: int64

This is date where orders were handled to the logistics partner. There are more order statuses that has missing values. 

609 orders were `unavailable`, which was unclear on what unavailable means in this context. 550 orders were cancelled, which probably included the 141 cancelled orders in the previous check. 314 `invoiced`, 301 `processing`, 5 `created`, and 1 `approved` would mean the orders havent been shipped (given to logistics) yet. 2 `delivered` was another anomaly, which could be caused by internal operations.

##### `order_delivered_customer_date`

In [20]:
df_orders["order_status"][df_orders["order_delivered_customer_date"].isna()].value_counts()

order_status
shipped        1107
canceled        619
unavailable     609
invoiced        314
processing      301
delivered         8
created           5
approved          2
Name: count, dtype: int64

This date is where orders were officially delivered to the customers by the logistics partner. 

The missing values for `shipped`, `canceled`, `unavailable`, `invoiced`, `processing`, `created, and `approved` all make sense as it means these orders are not yet delivered. The only anomaly, again, is the 8 orders that have `delivered` status but don't have `order_delivered_customer_date` value.

### Data Glimpse

In [28]:
df_orders[df_orders.isna().any(axis=1)]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
6,136cce7faa42fdb2cefd53fdc79a6098,ed0271e0b7da060a393796590e7b737a,invoiced,2017-04-11 12:22:08,2017-04-13 13:25:17,NaN,NaN,2017-05-09 00:00:00
44,ee64d42b8cf066f35eac1cf57de1aa85,caded193e8e47b8362864762a83db3c5,shipped,2018-06-04 16:44:48,2018-06-05 04:31:18,2018-06-05 14:32:00,NaN,2018-06-28 00:00:00
103,0760a852e4e9d89eb77bf631eaaf1c84,d2a79636084590b7465af8ab374a8cf5,invoiced,2018-08-03 17:44:42,2018-08-07 06:15:14,NaN,NaN,2018-08-21 00:00:00
128,15bed8e2fec7fdbadb186b57c46c92f2,f3f0e613e0bdb9c7cee75504f0f90679,processing,2017-09-03 14:22:03,2017-09-03 14:30:09,NaN,NaN,2017-10-03 00:00:00
154,6942b8da583c2f9957e990d028607019,52006a9383bf149a4fb24226b173106f,shipped,2018-01-10 11:33:07,2018-01-11 02:32:30,2018-01-11 19:39:23,NaN,2018-02-07 00:00:00
...,...,...,...,...,...,...,...,...
99283,3a3cddda5a7c27851bd96c3313412840,0b0d6095c5555fe083844281f6b093bb,canceled,2018-08-31 16:13:44,NaN,NaN,NaN,2018-10-01 00:00:00
99313,e9e64a17afa9653aacf2616d94c005b8,b4cd0522e632e481f8eaf766a2646e86,processing,2018-01-05 23:07:24,2018-01-09 07:18:05,NaN,NaN,2018-02-06 00:00:00
99347,a89abace0dcc01eeb267a9660b5ac126,2f0524a7b1b3845a1a57fcf3910c4333,canceled,2018-09-06 18:45:47,NaN,NaN,NaN,2018-09-27 00:00:00
99348,a69ba794cc7deb415c3e15a0a3877e69,726f0894b5becdf952ea537d5266e543,unavailable,2017-08-23 16:28:04,2017-08-28 15:44:47,NaN,NaN,2017-09-15 00:00:00
